state/pool.go

```diff
@@ -52,25 +59,51 @@ func (p *StatePool) Get(modelUUID string) (*State, error) {
 	if ok && item.remove {
 		// We don't want to allow increasing the refcount of a model
 		// that's been removed.
-		return nil, errors.Errorf("model %v has been removed", modelUUID)
+		return nil, nil, errors.Errorf("model %v has been removed", modelUUID)
 	}
+
+	p.soureKey++
+	key := p.soureKey
+	// released is here to be captured by the closure for the releaser.
+	// This is to ensure that the releaser function can only be called once.
+	released := false
+
+	releaser := func() {
+		if released {
+			return
+		}
+		err := p.release(modelUUID, key)
+		if err != nil {
+			logger.Errorf("releasing state back to pool: %s", err.Error())
+		}
+		released = true
+	}
+	source := string(debug.Stack())
```

Comment:
it is pretty quick, and certainly much faster than any access of mongo we'll be doing